src/function/memo.rs

```diff
@@ -1,108 +1,88 @@
+use std::any::Any;
+use std::fmt::Debug;
 use std::fmt::Formatter;
 use std::sync::Arc;
 
-use arc_swap::{ArcSwap, Guard};
 use crossbeam::atomic::AtomicCell;
 
+use crate::zalsa_local::QueryOrigin;
 use crate::{
-    hash::FxDashMap, key::DatabaseKeyIndex, zalsa::Zalsa, zalsa_local::QueryRevisions, Event,
-    EventKind, Id, Revision,
+    key::DatabaseKeyIndex, zalsa::Zalsa, zalsa_local::QueryRevisions, Event, EventKind, Id,
+    Revision,
 };
 
-use super::Configuration;
-
-/// The memo map maps from a key of type `K` to the memoized value for that `K`.
-/// The memoized value is a `Memo<V>` which contains, in addition to the value `V`,
-/// dependency information.
-pub(super) struct MemoMap<C: Configuration> {
-    map: FxDashMap<Id, ArcMemo<'static, C>>,
-}
+use super::{Configuration, IngredientImpl};
 
 #[allow(type_alias_bounds)]
-type ArcMemo<'lt, C: Configuration> = ArcSwap<Memo<<C as Configuration>::Output<'lt>>>;
+pub(super) type ArcMemo<'lt, C: Configuration> = Arc<Memo<<C as Configuration>::Output<'lt>>>;
```

Comment:
What's the remaining use case for `ArcMemo`? Aren't we now allocating the data in the pages or do the pages only store references to the arc?

Reply:
In this PR, the pages store a reference to the `Arc<Memo>`. I was debating about that, whether we could replace it with an `Id` of its own.

Reply:
Hmm, my understanding was that the main change in this PR is that memos are no longer stored in `Arc`s, instead they're stored directly in `Page`s to get an arena-like allocation. I remembered that it is important to you that all page-headers have the same layout. Is this the reason why we're keeping an `Arc` (or Id`) or are there other reasons for not storing the value in the pages? Or am I misunderstanding the change.

Reply:
Not exactly. I need to write up some docs on the layout. The big advantage in this PR though is that we do (on the fast path) two array lookups instead of a hash. But the memoized data is still stored in "arcs".